File: day1/aoc1.py
with open("aoc1_in.txt") as f:
    lines = f.read().split("\n")[:-1]

# day 1 puzzle 1
left = []
right = []
for line in lines:
    a, *b, c = line.split(" ")
    left.append(int(a))
    right.append(int(c))
left.sort()
right.sort()
distance = 0
for i in range(len(left)):
    distance += abs(left[i] - right[i])
print(distance)

# day 1 puzzle 2
right_map = {}
sim = 0
for loc in right:
    try:
        right_map[loc] = right_map[loc] + 1
    except KeyError:
        right_map[loc] = 1
for loc in left:
    try:
        sim = sim + (loc * right_map[loc])
    except KeyError:
        pass
print(sim)

File: day1/aoc1.rs
use std::fs;
use std::collections::HashMap;

fn main() {
    // day 1
    let filename = "aoc1_in.txt";
    let contents = fs::read_to_string(filename)
        .expect("file read err");
    let lines: Vec<&str> = contents.lines().collect();
    // println!("{:?}", lines);

    // puzzle 1
    let mut left: Vec<i32> = vec![];
    let mut right: Vec<i32> = vec![];
    for line in lines.iter() {
        let cols: Vec<&str> = line.split("   ").collect();
        let l: i32 = cols[0].parse().unwrap();
        let r: i32 = cols[1].parse().unwrap();
        left.push(l);
        right.push(r);
    }
    left.sort();
    right.sort();
    let mut distance: i32 = 0;
    for (i, l) in left.iter().enumerate() {
        let d = l - right[i];
        distance = distance + d.abs();
    }
    println!("{:?}", distance);

    // puzzle 2
    let mut counts = HashMap::new();
    for n in right.iter() {
        *counts.entry(n).or_insert(0) += 1;
    }
    let mut sim: i32 = 0;
    for l in left.iter() {
        sim = sim + (l * *counts.entry(l).or_insert(0));
    }
    println!("{:?}", sim);
}
